cluster/cluster.go

```diff
@@ -0,0 +1,83 @@
+package cluster
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"github.com/hashicorp/raft"
+	"github.com/narvikd/errorskit"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"log"
+	"nubedb/api/proto"
+	"nubedb/cluster/consensus/fsm"
+	"nubedb/internal/config"
+	"time"
+)
+
+func Execute(cfg config.Config, consensus *raft.Raft, payload *fsm.Payload) error {
+	payloadData, errMarshal := json.Marshal(&payload)
+	if errMarshal != nil {
+		return errorskit.Wrap(errMarshal, "couldn't marshal data to send it to the DB cluster")
+	}
+
+	if consensus.State() != raft.Leader {
+		return handleForwardLeaderFuture(cfg, consensus, payload)
+	}
+
+	return applyLeaderFuture(consensus, payloadData)
+}
+
+func applyLeaderFuture(consensus *raft.Raft, payloadData []byte) error {
+	const timeout = 500 * time.Millisecond
+
+	future := consensus.Apply(payloadData, timeout)
+	if future.Error() != nil {
+		return errorskit.Wrap(future.Error(), "couldn't persist data to DB Cluster")
+	}
+
+	response := future.Response().(*fsm.ApplyRes)
+	if response.Error != nil {
+		return errorskit.Wrap(response.Error, "DB cluster returned an error when trying to persist data to it")
+	}
+
+	return nil
+}
+
+func handleForwardLeaderFuture(cfg config.Config, consensus *raft.Raft, payload *fsm.Payload) error {
+	_, leaderID := consensus.LeaderWithID()
+	leaderCfg := cfg.Nodes[string(leaderID)]
+	return forwardLeaderFuture(leaderCfg, payload)
+}
+
+func forwardLeaderFuture(leaderCfg config.NodeCfg, payload *fsm.Payload) error {
```

Comment:
Where's the timeout in this func?

Reply:
Done: https://github.com/narvikd/nubedb/pull/10/commits/7435836e2357026fe9927087c3e9b3961dcdd89c